test/conviction-voting.js

```diff
@@ -31,367 +43,648 @@ function calculateThreshold(requested, funds, supply, alpha, beta, rho) {
+            await convictionVoting.executeProposal(proposalId)
+            const { convictionLast: convictionBefore, blockLast: blockLastBefore } = await convictionVoting.getProposal(proposalId)
+
+            await convictionVoting.withdrawFromProposal(proposalId, 1000, { from: user })
+
+            const { convictionLast: convictionAfter, blockLast: blockLastAfter } = await convictionVoting.getProposal(proposalId)
+            assert.equal(blockLastAfter.toString(), blockLastBefore.toString(), 'Incorrect block last')
+            assert.equal(convictionAfter.toString(), convictionBefore.toString(), 'Incorrect conviction')
+          })
+
+          it('reverts when withdrawing more than staked', async () => {
+            await assertRevert(convictionVoting.withdrawFromProposal(proposalId, stakeAmount + 1), 'CV_WITHDRAW_MORE_THAN_STAKED')
+          })
+
+          it('reverts when withdrawing 0', async () => {
+            await assertRevert(convictionVoting.withdrawFromProposal(proposalId, 0), 'CV_AMOUNT_CAN_NOT_BE_ZERO')
           })
-          assert.equal(
-            (await app.getProposal(i))[3].toNumber(),
-            account === appManager ? 0 : 7458,
-            'Conviction does not match expectations'
-          )
-          const totalTokensStaked = getEventArgument(
-            receipt,
-            'StakeChanged',
-            'totalTokensStaked'
-          ).toNumber()
-          assert.equal(totalTokensStaked, currentlyStaked + stakesPerAccount)
-          const tokensStaked = getEventArgument(
-            receipt,
-            'StakeChanged',
-            'tokensStaked'
-          ).toNumber()
-          assert.equal(tokensStaked, stakesPerAccount)
         })
-      }
 
-      it(`should withdraw from proposal - by ${acc[appManager]}`, async () => {
-        // assume that 1 block ~ 15 seconds
-        // +20 blocks
-        await timeAdvancer.advanceTimeAndBlocksBy(15 * 20, 20)
+        context('withdrawAllFromProposal(uint256 _proposalId)', () => {
+          it('withdraws all from proposal', async () => {
+            const stakeAmount = 1000
+            await convictionVoting.stakeToProposal(proposalId, stakeAmount)
+            const { stakedTokens: stakedBefore } = await convictionVoting.getProposal(proposalId)
+            const currentBlock = await convictionVoting.getBlockNumberPublic()
 
-        const currentlProposal = await app.getProposal(i)
-        const currentlyStaked = currentlProposal[2]
-        const withdrawAmount = 1000
+            await convictionVoting.withdrawAllFromProposal(proposalId)
 
-        const receipt = await app.withdrawFromProposal(i, withdrawAmount, {
-          from: appManager,
+            await assertProposalAndStakesCorrect(
+              proposalId, 1000, stakedBefore.toNumber() - stakeAmount,
+              currentBlock.toNumber() + 1, 0, 0, [], 0)
+          })
+        })
+
+        context('_onTransfer(address _from, address _to, uint256 _amount)', () => {
```

Comment:
Nicely done on these tests!

Consider also adding one for when the user is staking on the ABSTAIN proposal